lib/Service/ParticipantService.php

```diff
@@ -361,6 +363,77 @@ public function addGroup(Room $room, IGroup $group, array $existingParticipants
 		$this->addUsers($room, $newParticipants);
 	}
 
+	/**
+	 * @param Room $room
+	 * @param Circle $circle
+	 * @param Participant[] $existingParticipants
+	 */
+	public function addCircle(Room $room, Circle $circle, array $existingParticipants = []): void {
+		$membersInCircle = $circle->getMembers();
+
+		if (empty($existingParticipants)) {
+			$existingParticipants = $this->getParticipantsForRoom($room);
+		}
+
+		$participantsByUserId = [];
+		foreach ($existingParticipants as $participant) {
+			if ($participant->getAttendee()->getActorType() === Attendee::ACTOR_USERS) {
+				$participantsByUserId[$participant->getAttendee()->getActorId()] = $participant;
+			}
+		}
+
+		$newParticipants = [];
+		foreach ($membersInCircle as $member) {
+			/** @var Member $member */
+			if ($member->getUserType() !== Member::TYPE_USER || $member->getUserId() === '') {
+				// Not a user?
+				continue;
+			}
+
+			if ($member->getStatus() !== Member::STATUS_INVITED && $member->getStatus() !== Member::STATUS_MEMBER) {
+				// Only allow invited and regular members
+				continue;
+			}
+
+			$user = $this->userManager->get($member->getUserId());
+			if (!$user instanceof IUser) {
+				continue;
+			}
+
+			$existingParticipant = $participantsByUserId[$user->getUID()] ?? null;
+			if ($existingParticipant instanceof Participant) {
+				if ($existingParticipant->getAttendee()->getParticipantType() === Participant::USER_SELF_JOINED) {
+					$this->updateParticipantType($room, $existingParticipant, Participant::USER);
+				}
+
+				// Participant is already in the conversation, so skip them.
+				continue;
+			}
+
+			$newParticipants[] = [
+				'actorType' => Attendee::ACTOR_USERS,
+				'actorId' => $user->getUID(),
+				'displayName' => $user->getDisplayName(),
+			];
+		}
+
+		// No tracking of circles for now
+//		try {
+//			$this->attendeeMapper->findByActor($room->getId(), Attendee::ACTOR_CIRCLES, $circle->getSingleId());
+//		} catch (DoesNotExistException $e) {
+//			$attendee = new Attendee();
+//			$attendee->setRoomId($room->getId());
+//			$attendee->setActorType(Attendee::ACTOR_CIRCLES);
+//			$attendee->setActorId($circle->getSingleId());
+//			$attendee->setDisplayName($circle->getDisplayName());
+//			$attendee->setParticipantType(Participant::USER);
+//			$attendee->setReadPrivacy(Participant::PRIVACY_PRIVATE);
+//			$this->attendeeMapper->insert($attendee);
```

Comment:
I didn't find this code in the non-refactored version so I assume this is a "would-be in the future" that you've just added here.

to me, without doing a line by line compare, it looks like this PR only contains refactoring but no actual behavior change, since the above is commented out

can you point to the workaround/fix for the getType issue ? (or split the PR into one commit for refactoring and one for the fix)

Reply:
https://github.com/nextcloud/spreed/pull/5691/files#diff-d583983624ef8769460d4aa59e64b75ee54b6955e01e620e3eacedb7ba208786L717
and
https://github.com/nextcloud/spreed/pull/5691/files#diff-d583983624ef8769460d4aa59e64b75ee54b6955e01e620e3eacedb7ba208786L1062
both used getType

The fixed method uses getUserType:
https://github.com/nextcloud/spreed/pull/5691/files#diff-8d29629ca6909ff3b00eb3ff13094b704d2b10ce22eff12e6921dccf08a83100R388

The commented out code from here is basically from #5670 and because this function is basically a close of addGroup() right above.

Reply:
thanks for the explanation, makes sense